Link detection for the exception widget and debug console

michelkaporin opened the pull request:
1. Extracted link detection to a separate common class, shared and reused between REPL viewer and exception widget. 
2. Updated link detection to handle multiple links in a provided string.
3. Added link validation to resolve only files which are part of file system.

Partially reused regex of the reverted PR by @felixfbecker, credits to you!

![image](https://cloud.githubusercontent.com/assets/2239563/24871735/ef9965ea-1e1a-11e7-8c5d-e01d915a4c44.png)
![image](https://cloud.githubusercontent.com/assets/2239563/24871911/9f107a04-1e1b-11e7-9278-ffe90b694cbe.png)

Fixes #15760, #17085 (and all its ancestors).
Partially implements #21349.

felixfbecker commented:
Awesome! Did you test this with stack traces printed by mocha?

michelkaporin commented:
@felixfbecker Thank you :) I did not explicitly test mocha but those mocha stack traces described in #15760 should be picked up now (see the the sample image above which is a super set of what was outputted there).

The things that won't work currently:
1. If file is coming from <node internals> or does not exist on file system, it won't be anchored.
2. http:// links still not supported.
3. If the path to the file defined without the column, it won't be recognised.

I plan to tackle those things in a separate issues.

felixfbecker commented:
I don't see this explicitely in the image above, will it work for relative paths that do not take up a whole line? Eg
```
  at myfunc (src/myfunc.ts:1:0)
```
right now that only highlights `/myfunc.ts:1:0` like it was an absolute path

michelkaporin commented:
@felixfbecker yes, it will work for relative paths that do not take up a whole line. `src/myfunc.ts:1:0` would be highlighted in your example. 

![image](https://cloud.githubusercontent.com/assets/2239563/24896780/bff0f070-1e96-11e7-9bc9-614ec22f6237.png)

isidorn commented:
@michelkaporin I appologize for the slow response was busy around the release.

Just checked the PR, looks great!
I really like that you are reusing the logic between the repl and the `exceptionWidget`.

After discussing with @bpasero we agreed that doing a FS stat to figure out if something is a relative link is too heavy for a simple operation like this.

How precise would the relative path mathching be if you can not verify it against the real file system? If not very precise then we should not support relative file matching.
But all in all we need to drop the dependency to `'fs'` in the `linkDetector`

felixfbecker commented:
I think it should be doable to detect relative links without a stat. If spaces in paths are not to be supported, it's trivial. Otherwise you would need to have to look at common seperators like parenthesis. But please please keep relative link detection, I need to click on these >100 times a day in stack traces.

michelkaporin commented:
@isidorn Ok, I can drop link validation and try to match it with regex then. 
However, current terminal implementation is also doing `fs.stat` to validate the links in the terminal. If we plan to reuse link detection code between terminal, debug console and exception widget, then we will have to drop `fs.stat` for terminal as well.

isidorn commented:
Thanks for letting me know, we should also not use `fs.stat` for link detection in the terminal.
fyi @Tyriar

Tyriar commented:
@michelkaporin @isidorn @bpasero I disagree that we shouldn't stat the files. Without `fs.stat` there are a heap of false positives. A lot of stuff in the terminal is formed like a path, especially if we want to linkify files in the output of `ls`. Note that the terminal does not linkify all output, it happens 200ms after the viewport has not changed, so it will stat as many links that are matched in the viewport (typically 10-20 lines) at most every 200ms.

Related:

- "Link validation" introduced in 1.11 (ie. stat the paths) https://code.visualstudio.com/updates/v1_11#_link-improvements, it was pretty bad before this was added imo.
- I brought up the concern when linkifying `ls` output and suggested we opt for `readdir` there instead. https://github.com/Microsoft/vscode/issues/22772#issuecomment-293625442
- xterm.js link impl https://github.com/sourcelair/xterm.js/blob/master/src/Linkifier.ts

bpasero commented:
Stating things we "think" are links is crazy to put it in simple words. Since output is unbounded this could easily cause millions of disk access for no helpful reason. A link should be detected without going to disk and it is totally fine to have false positives (e.g. it is OK to not be 100% correct). Once the user clicks on a link, it is fine to show an error if the file does not exist.

Tyriar commented:
@bpasero to be more specific the terminal stuff [is using `pfs`](https://github.com/Microsoft/vscode/blob/44e6c6dcfcd72f291be00f42d20a8f4fd7c69121/src/vs/workbench/parts/terminal/electron-browser/terminalLinkHandler.ts#L192) so it's not a blocking call or anything. Also for the terminal it *is* bounded as only the current viewport is checked for links.

Is it so bad to check the existence of a handful of paths after a terminal command is run (or when the user scrolls the terminal)?

bpasero commented:
I would not do it, the output panel has never done it. There are tons of regexes that try to detect links and render them optimistically. What is wrong with that approach?

Tyriar commented:
Tracking stat in terminal links in https://github.com/Microsoft/vscode/issues/25024

michelkaporin commented:
@isidorn I have updated regex to be more precise without `fs.stat` validation. We get higher false positive rate now without it, although it is more efficient. 

Exception widget is an example of high false positive rate with file names coming from debug adapter, that are detected by regex as workspace' relative links, whereas they are not, in reality. On the good side the user will get an 'Unable to open' message, when clicking on such links.

isidorn commented on src/vs/workbench/parts/debug/browser/linkDetector.ts:
```diff
@@ -0,0 +1,107 @@
+		// group 3: line number
+		// group 4: column number
+		// eg: at Context.<anonymous> (c:\Users\someone\Desktop\mocha-runner\test\test.js:26:11)
+		/(?![\(])(?:file:\/\/)?((?:([a-zA-Z]+:)|[^\(\)<>\'\"\[\]:\s]+)(?:[\\/][^\(\)<>\'\"\[\]:]*)?\.[a-zA-Z]+[0-9]*):(\d+)(?::(\d+))?/g
```
I notice that the comments are the same but the regex is now more complicated.
Could you please add some comments in code to clarify this regex beast

michelkaporin replied:
@isidorn I've added more comments, hope it makes things more clear. It is quite big just because it tries **not** to match lots of special characters. In the end we still plan to move away from it to the common regex between terminal, debug console and exception widget.

isidorn commented on src/vs/workbench/parts/debug/browser/linkDetector.ts:
```diff
@@ -0,0 +1,107 @@
+		// noop
+	}
+
+	public handleLinks(text: string): HTMLElement | string {
```
As far as I understand there have been no changes till the end of this file, it has just been moved over.
If that is the case it looks good to me

michelkaporin replied:
It was changed to handle multiple links in the provided string. Previously it used to detect only first link match and then stopped.

isidorn commented on src/vs/workbench/parts/debug/browser/linkDetector.ts:
```diff
@@ -0,0 +1,107 @@
+	) {
+		// noop
+	}
+
```
Since this handleLinks is now a public method we could try to make the API a bit nicer, a personally find it ugly that it returns an HTMLElement or a string. What does that mean?
At least provide some comments to clarify each case, or try to always return an object with the corresponding fields.

michelkaporin replied:
I've added the comment to it. I hope this explains it well. 

If you have any suggestions on what should be a return result instead, I am happy to hear. Currently it perfectly suits exception widget and debug console needs.

isidorn commented on src/vs/workbench/parts/debug/browser/media/exceptionWidget.css:
```diff
@@ -27,6 +27,11 @@
 	margin-top: 0.5em;
 }
 
```
Makes sense

isidorn commented on src/vs/workbench/parts/debug/electron-browser/replViewer.ts:
```diff
@@ -7,13 +7,10 @@ import * as nls from 'vs/nls';
 import { TPromise } from 'vs/base/common/winjs.base';
```
Looks good overall and I like we removed code from this class.

I only have the issue that you are creating a `linkDetector` at every iteration in a loop. Would it not make sense just to create one global one for the replViewer and always to reuse it?

michelkaporin replied:
@isidorn You're right, did not see that for-loop when refactored. I've moved LinkDetector instantiation  to the constructor.

isidorn commented on src/vs/workbench/parts/debug/browser/exceptionWidget.ts:
```diff
@@ -11,9 +11,11 @@ import { ICodeEditor } from 'vs/editor/browser/editorBrowser';
 import { IContextViewService } from 'vs/platform/contextview/browser/contextView';
```
looks good

isidorn commented:
@michelkaporin great work. I have added some comments directly in the code. Once you adrress them we can merge it in.
I see that the builds are failing, is everything ok when you build it localy?

michelkaporin commented:
@isidorn thank you! I've addressed all your comments, please see my replies. The build is passing locally, AppVeyour succeeds also on the latest codebase. Cannot find any build issues in the Travis log that are to do with my changes.

isidorn commented:
@michelkaporin looks good, let's merge it in to see how it behaves :)
